# Lecture 10 Data Manipulation with `pandas`

- [10.1 Introduction to pandas](#section1)
- [10.2 Importing Data and Summary Statistics](#section2)
- [10.3 Rename, Index, and Slice](#section3)
- [10.4 Creating New Columns, Reordering](#section4)
- [10.5 Merging](#section5)
- [10.6 Calculating Unique and Missing Values](#section6)
- [10.7 Exporting A DataFrame to csv](#section7)
- [10.8 Dealing With Missing Values: Boolean Indexing](#section8)
- [References](#section9)


# 10.1 Introduction to pandas <a id="section1"/>

`pandas` is a library designed for working with structured data, such as tabular data (e.g., from .csv files, Excel files) or relational databases (e.g., SQL). 

> The ***DataFrame*** object in `pandas` is a 2-dimensional tabular, column-oriented data structure. The DateFrame is similar to a spreadsheet and can store data of different types (including text characters, integers, floating-point values, categorical data, and more).

<img style="float: center; height:260px;" src="images/pic1.jpg">

The `pandas` name is derived from the term *panel data*, which is a term from economics for multi-dimensional structured data. 

For more information about `pandas` you can consult the full [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/).

In this lecture, we will first work with example data and go through the steps of preparing data for analysis. Using `pandas` for these tasks is much more convenient than using Python's built-in data structures.

# 10.2 Importing Data and Summary Statistics <a id="section2"/>

Let's begin by importing the `pandas` package using the common abbreviation `pd`. Loading `pandas` as `pd` is standard practice.

In [1]:
import pandas as pd

A wide range of input/output formats are supported by `pandas`:

* CSV, text
* SQL database
* Excel
* HDF5
* json
* html
* pickle
* sas, stata
* ...

For importing .csv files, the `read_csv()` function in `pandas` allows us to easily import data. By default, it assumes that the data is comma-separated, but we can also specify the delimiter used in the data (e.g., tab, semicolon, etc.). There are several parameters that can be specified in `read_csv()`. See the documentation [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). 

Let's load the data in the file `country-total` located in the folder `data` and save it under the name `unemployment`. This file contains unemployment information for several countries over a time period.

The function `read_csv()` returns a `DataFrame`, as shown below. Note that the data has 20,796 rows, and the output of the cell displayed only the first 5 and last 5 rows (or the first and last 30, depending on your system), since displaying all 20,976 rows is probably not what we want at this point anyway.

In [2]:
# Import data
unemployment = pd.read_csv('data/country_total.csv')
# Show the DataFrame
unemployment

country seasonality    month  unemployment  unemployment_rate
0          at         nsa  1993.01        171000                4.5
1          at         nsa  1993.02        175000                4.6
2          at         nsa  1993.03        166000                4.4
3          at         nsa  1993.04        157000                4.1
4          at         nsa  1993.05        147000                3.9
5          at         nsa  1993.06        134000                3.5
6          at         nsa  1993.07        128000                3.4
7          at         nsa  1993.08        130000                3.4
8          at         nsa  1993.09        132000                3.5
9          at         nsa  1993.10        141000                3.7
10         at         nsa  1993.11        156000                4.1
11         at         nsa  1993.12        169000                4.4
12         at         nsa  1994.01        180000                4.7
13         at         nsa  1994.02        180000                4.7
14         at         nsa  1994.03        164000                4.3
15         at         nsa  1994.04        152000                4.0
16         at         nsa  1994.05        143000                3.8
17         at         nsa  1994.06        127000                3.4
18         at         nsa  1994.07        120000                3.2
19         at         nsa  1994.08        121000                3.2
20         at         nsa  1994.09        125000                3.3
21         at         nsa  1994.10        134000                3.5
22         at         nsa  1994.11        148000                3.9
23         at         nsa  1994.12        159000                4.2
24         at         nsa  1995.01        172000                4.5
25         at         nsa  1995.02        174000                4.5
26         at         nsa  1995.03        160000                4.2
27         at         nsa  1995.04        150000                3.9
28         at         nsa  1995.05        141000                3.7
29         at         nsa  1995.06        129000                3.4
...       ...         ...      ...           ...                ...
20766      uk       trend  2008.05       1667000                5.4
20767      uk       trend  2008.06       1712000                5.5
20768      uk       trend  2008.07       1774000                5.7
20769      uk       trend  2008.08       1822000                5.9
20770      uk       trend  2008.09       1859000                6.0
20771      uk       trend  2008.10       1914000                6.1
20772      uk       trend  2008.11       1976000                6.3
20773      uk       trend  2008.12       2038000                6.5
20774      uk       trend  2009.01       2120000                6.8
20775      uk       trend  2009.02       2199000                7.0
20776      uk       trend  2009.03       2264000                7.2
20777      uk       trend  2009.04       2341000                7.5
20778      uk       trend  2009.05       2408000                7.7
20779      uk       trend  2009.06       2440000                7.8
20780      uk       trend  2009.07       2441000                7.8
20781      uk       trend  2009.08       2441000                7.8
20782      uk       trend  2009.09       2444000                7.8
20783      uk       trend  2009.10       2437000                7.8
20784      uk       trend  2009.11       2423000                7.8
20785      uk       trend  2009.12       2423000                7.8
20786      uk       trend  2010.01       2456000                7.9
20787      uk       trend  2010.02       2463000                7.9
20788      uk       trend  2010.03       2437000                7.8
20789      uk       trend  2010.04       2419000                7.8
20790      uk       trend  2010.05       2419000                7.7
20791      uk       trend  2010.06       2429000                7.7
20792      uk       trend  2010.07       2422000              

We could have also used print to display the DataFrame.

In [3]:
print(unemployment)

      country seasonality    month  unemployment  unemployment_rate
0          at         nsa  1993.01        171000                4.5
1          at         nsa  1993.02        175000                4.6
2          at         nsa  1993.03        166000                4.4
3          at         nsa  1993.04        157000                4.1
4          at         nsa  1993.05        147000                3.9
5          at         nsa  1993.06        134000                3.5
6          at         nsa  1993.07        128000                3.4
7          at         nsa  1993.08        130000                3.4
8          at         nsa  1993.09        132000                3.5
9          at         nsa  1993.10        141000                3.7
10         at         nsa  1993.11        156000                4.1
11         at         nsa  1993.12        169000                4.4
12         at         nsa  1994.01        180000                4.7
13         at         nsa  1994.02        180000

When the DataFrames have a large number of rows that take large portion of the screen, we can inspect the data by using the `.head()` method. By default, this shows the ***header*** (names of the columns, commonly referred to as ***column labels***) and the first five rows (having indices ranging from 0 to 4, in the first column in the table). The ***indices*** are also referred to as ***row labels***.

In [4]:
unemployment.head()

country seasonality    month  unemployment  unemployment_rate
0      at         nsa  1993.01        171000                4.5
1      at         nsa  1993.02        175000                4.6
2      at         nsa  1993.03        166000                4.4
3      at         nsa  1993.04        157000                4.1
4      at         nsa  1993.05        147000                3.9

Passing an integer number as an argument to `.head(n)` returns that number of rows. To see the last $n$ rows, use `.tail(n)`.

In [5]:
# show the last 8 rows
unemployment.tail(8)

country seasonality    month  unemployment  unemployment_rate
20788      uk       trend  2010.03       2437000                7.8
20789      uk       trend  2010.04       2419000                7.8
20790      uk       trend  2010.05       2419000                7.7
20791      uk       trend  2010.06       2429000                7.7
20792      uk       trend  2010.07       2422000                7.7
20793      uk       trend  2010.08       2429000                7.7
20794      uk       trend  2010.09       2447000                7.8
20795      uk       trend  2010.10       2455000                7.8

To find the number of rows in a DataFrame, you can use the `len()` function, as with Python lists and other sequences.

In [6]:
len(unemployment)

20796

Alternatively, you can use the `shape` attribute to find the numbers of rows and columns, as with NumPy arrays. The cell output is a tuple, showing that there are 20,796 rows and 5 columns. Note that the left-most column in the above table showing row indices is not part of the data.

In [7]:
unemployment.shape

(20796, 5)

A useful method that generates various summary statistics of a DataFrame is `.describe()`, as shown below.

You can notice in the above cell that the data has 5 columns, but the first 2 columns (country and seasonality) have textual (strings) data, and therefore the summary statistics are shown only for the columns with numeric data (month, unemployment, and unemployment_rate). If `.describe()` is called on textual data only, it will return the count, number of unique values, and the most frequent value along with its count.

In [8]:
unemployment.describe()

month  unemployment  unemployment_rate
count  20796.000000  2.079600e+04       19851.000000
mean    1999.401290  7.900818e+05           8.179764
std        7.483751  1.015280e+06           3.922533
min     1983.010000  2.000000e+03           1.100000
25%     1994.090000  1.400000e+05           5.200000
50%     2001.010000  3.100000e+05           7.600000
75%     2006.010000  1.262250e+06          10.000000
max     2010.120000  4.773000e+06          20.900000

It is also possible to calculate individual statistics, such as `.min()`, `.max ()`, or `.mean()`, instead of using summary statistics with `.describe()`.

In [9]:
unemployment.min()

country                   at
seasonality              nsa
month                1983.01
unemployment            2000
unemployment_rate        1.1
dtype: object

To view the data types for each column data use the `dtypes` attribute of the *unemployment*. The data types in this case are strings (object), floats (float64), and integers (int64).

In [10]:
unemployment.dtypes

country               object
seasonality           object
month                float64
unemployment           int64
unemployment_rate    float64
dtype: object

And one more way to get a summary of the data is by using `info()`.

In [11]:
unemployment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20796 entries, 0 to 20795
Data columns (total 5 columns):
country              20796 non-null object
seasonality          20796 non-null object
month                20796 non-null float64
unemployment         20796 non-null int64
unemployment_rate    19851 non-null float64
dtypes: float64(2), int64(1), object(2)
memory usage: 812.4+ KB


### Import Data From a URL

Above, we imported the unemployment data using the `read_csv` function and a relative file path to the `data` directory. `read_csv` is a very flexible method and it also allows importing data using a URL as the file path. 

A csv file with data on world countries and their abbreviations is located at [https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv](https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv)

Using `read_csv`, we can import the country data and save it to the variable `countries`. This DataFrame has 30 rows.

In [12]:
countries = pd.read_csv('https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv'                       )
countries

country google_country_code country_group  \
0       at                  AT            eu   
1       be                  BE            eu   
2       bg                  BG            eu   
3       hr                  HR        non-eu   
4       cy                  CY            eu   
5       cz                  CZ            eu   
6       dk                  DK            eu   
7       ee                  EE            eu   
8       fi                  FI            eu   
9       fr                  FR            eu   
10      de                  DE            eu   
11      gr                  GR            eu   
12      hu                  HU            eu   
13      ie                  IE            eu   
14      it                  IT            eu   
15      lv                  LV            eu   
16      lt                  LT            eu   
17      lu                  LU            eu   
18      mt                  MT            eu   
19      nl                  NL            eu   
20      no                  NO        non-eu   
21      pl                  PL            eu   
22      pt                  PT            eu   
23      ro                  RO            eu   
24      sk                  SK            eu   
25      si                  SI            eu   
26      es                  ES            eu   
27      se                  SE            eu   
28      tr                  TR        non-eu   
29      uk                  GB            eu   

                                      name_en  \
0                                     Austria   
1                                     Belgium   
2                                    Bulgaria   
3                                     Croatia   
4                                      Cyprus   
5                              Czech Republic   
6                                     Denmark   
7                                     Estonia   
8                                     Finland   
9                                      France   
10  Germany (including  former GDR from 1991)   
11                                     Greece   
12                                    Hungary   
13                                    Ireland   
14                                      Italy   
15                                     Latvia   
16                                  Lithuania   
17                                 Luxembourg   
18                                      Malta   
19                                Netherlands   
20                                     Norway   
21                                     Poland   
22                                   Portugal   
23                                    Romania   
24                                   Slovakia   
25                                   Slovenia   
26                                      Spain   
27                                     Sweden   
28                                     Turkey   
29                             United Kingdom   

                                              name_fr  \
0                                            Autriche   
1                                            Belgique   
2                                            Bulgarie   
3                                             Croatie   
4                                              Chypre   
5                                  République tchèque   
6                                            Danemark   
7                                             Estonie   
8                                            Finlande   
9                                              France   
10  Allemagne (incluant l'ancienne RDA à partir de...   
11                                              Grèce   
12                                            Hongrie   
13                                            Irlande   
14                                             Italie   
15                                           Lettonie   
16                       

Similar to the above example, we can use `shape` and `describe()` the understand the *countries* DataFrame. In this case `.describe()` is not very useful, because only 2 of the columns have numeric values.

In [13]:
countries.shape

(30, 8)

In [14]:
# explore the countries data
countries.describe()

latitude  longitude
count  30.000000  30.000000
mean   49.092609  14.324579
std     7.956624  11.257010
min    35.129141  -8.239122
25%    43.230916   6.979186
50%    49.238087  14.941462
75%    54.090400  23.351690
max    64.950159  35.439795

### Import Data from Excel File

In a similar way, we can import data from an Excel file using the `read_excel()` function.

In [15]:
titanic = pd.read_excel('data/titanic.xls')
titanic

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen "Nellie"   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex     age  sibsp  parch    ticket      fare    cabin embarked  boat  \
0   female      29      0      0     24160  211.3375       B5        S     2   
1     male  0.9167      1      2    113781  151.5500  C22 C26        S    11   
2   female       2      1      2    113781  151.5500  C22 C26        S  None   
3     male      30      1      2    113781  151.5500  C22 C26        S  None   
4   female      25      1      2    113781  151.5500  C22 C26        S  None   
5     male      48      0      0     19952   26.5500      E12        S     3   
6   female      63      1      0     13502   77.9583       D7        S    10   
7     male      39      0      0    112050    0.0000      A36        S  None   
8   female      53      2      0     11769   51.4792     C101        S     D   
9     male      71      0      0  PC 17609   49.5042     None        C  None   
10    male      47      1      0  PC 17757  227.5250  C62 C64        C  None   
11  female      18      1      0  PC 17757  227.5250  C62 C64        C     4   
12  female      24      0      0  PC 17477   69.3000      B35        C     9   
13  female      26      0      0     19877   78.8500     None        S     6   
14    male      80      0      0     27042   30.0000      A23        S     B   
15    male    None      0      0  PC 17318   25.9250     None        S  None   
16    male      24      0      1  PC 17558  247.5208  B58 B60        C  None   
17  female      50      0      1  PC 17558  247.5208  B58 B60        C     6   
18  female      32      0      0     11813   76.2917      D15        C     8   
19    male      36      0      0     13050   75.2417       C6        C     A   

    body                        home.dest  
0   None                     St Louis, MO  
1   None  Montreal, PQ / Chesterville, ON  
2   None  Montreal, PQ / Chesterville, ON  
3    135  Montreal, PQ / Chesterville, ON  
4   None  Montreal, PQ / Chesterville, ON  
5   None                     New York, NY  
6   None                       Hudson, NY  
7   None                      Belfast, NI  
8   None              Bayside, Queens, NY  
9     22              Montevideo, Uruguay  
10   124                     New York, NY  
11  None                     New York, NY  
12  None                    Paris, France  
13  None                             None  
14  None                    Hessle, Yorks  
15  None                     New York, NY  

Note that all DataFrames that we loaded so far, i.e., `unemployment`, `countries`, and `titanic`, are stored in the memory, and we can access them when needed once they are loaded. For example:

In [16]:
titanic.shape

(20, 14)

### Create a DataFrame

Alternatively, we can manually create DataFrames, instead of importing from a file. The following DataFrame called *simple_table* contains information from the Titanic dataset. You can notice that the DataFrame is created similarly to creating a dictionary, where the column headers represent keys, and the data in each column are lists of values.

In [17]:
simple_table = pd.DataFrame({
        "Name": ["Braund, Mr. Owen Harris",
                 "Allen, Mr. William Henry",
                 "Bonnell, Miss. Elizabeth"],
        "Age": [22, 35, 58],
        "Sex": ["male", "male", "female"]})
simple_table

Name  Age     Sex
0   Braund, Mr. Owen Harris   22    male
1  Allen, Mr. William Henry   35    male
2  Bonnell, Miss. Elizabeth   58  female

In [18]:
simple_table.shape

(3, 3)

# 10.3 Rename, Index, and Slice <a id="section3"/>

Let's look again at the *unemployment* data. You may have noticed that the `month` column actually includes the year and the months added as decimals (e.g., 1993.01 should be year 1993 and month 01).

In [19]:
unemployment.head(3)

country seasonality    month  unemployment  unemployment_rate
0      at         nsa  1993.01        171000                4.5
1      at         nsa  1993.02        175000                4.6
2      at         nsa  1993.03        166000                4.4

Let's rename the column into *year_month*. The `.rename()` method allows modifying column and/or row names. As you can see in the cell below, we passed a dictionary to the `columns` parameter, with the original name `mont` as the key and the new name `year_month` as the value. Importantly, we also set the `inplace` parameter to `True`, which indicates that we want to modify the *actual* DataFrame, and not create a new copy of it.

In [20]:
unemployment.rename(columns={'month' : 'year_month'}, inplace=True)
unemployment.head(3)

country seasonality  year_month  unemployment  unemployment_rate
0      at         nsa     1993.01        171000                4.5
1      at         nsa     1993.02        175000                4.6
2      at         nsa     1993.03        166000                4.4

### Selecting Columns

To select a single column of the DataFrame, we can either use the name of the column enclosed in square brackets `[]` or the dot notation (referred to as *attribute access*). It is preferable to use the square brackets notation, since a column name might inadvertently have the same name as an built-in pandas method.

In [21]:
unemployment['year_month'].head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

In [22]:
unemployment.year_month.head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

When selecting a single column, we obtain a `pandas` **Series** object, which is a single vector of data with an associated array of index row labels shown in the left-most column.

> A ***Series*** object in `pandas` represents an 1-dimensional vector of data (i.e., a column of data).

If we check the type of the `unemployment` object and `unemployment['year_month']` we can see that the first one is DataFrame and the second one is Series.

In [23]:
type(unemployment)

pandas.core.frame.DataFrame

In [24]:
type(unemployment['year_month'])

pandas.core.series.Series

`pandas` provide many methods that can be applied to `Series` objects. A few examples are shown below.

In [25]:
print('minimum is ', unemployment['year_month'].min())
print('maximum is ', unemployment['year_month'].max())
print('mean value is ', unemployment['year_month'].mean())

minimum is  1983.01
maximum is  2010.12
mean value is  1999.4012896711013


To select multiple columns, use a list of column names within the selection brackets `[]`.

In [26]:
unemployment[['country','year_month']].head()

country  year_month
0      at     1993.01
1      at     1993.02
2      at     1993.03
3      at     1993.04
4      at     1993.05

### Selecting Rows

Another graphical representation of a DataFrame is shown in the figure below.

<img style="float: center; height:320px;" src="images/dataframe.png">

One way to select rows is by using the `[]` operator, similar to indexing and slicing in Python lists and other sequence data.

In [27]:
unemployment[0:4]

country seasonality  year_month  unemployment  unemployment_rate
0      at         nsa     1993.01        171000                4.5
1      at         nsa     1993.02        175000                4.6
2      at         nsa     1993.03        166000                4.4
3      at         nsa     1993.04        157000                4.1

The first column with the indices in `pandas` DataFrames do not need to be a sequence of integers, but they can also be strings or other numeric data (e.g., dates, years).

For instance, let's create a `DataFrame` called *bacteria* to see how indexing with string indices works. We again pass in a dictionary, with the keys corresponding to column names and the values to the data, and in addition we pass a list of strings called `index`. (Compare to the *simple_table* above, which does not use `index` for creating the DataFrame, and the indices are automatically set to integer numbers.)

In [28]:
bacteria = pd.DataFrame({'bacteria_counts' : [632, 1638, 569, 115],
                         'other_feature' : [438, 833, 234, 298]},
                         index=['Firmicutes', 'Proteobacteria', 'Actinobacteria', 'Bacteroidetes'])
bacteria

bacteria_counts  other_feature
Firmicutes                  632            438
Proteobacteria             1638            833
Actinobacteria              569            234
Bacteroidetes               115            298

In `pandas`, two other operators are used for selecting rows and/or columns: `.loc` and `.iloc`. 

`.loc` works with string labels, and `.iloc` works with integer labels. These two operators can accept a single label, a list of labels, or a slice of labels (e.g., `'a:f'` or `2:5`).

If we're interested in the row *Actinobacteria*, we can use `.loc` and the index name. This returns the column values for the specified row.

In [29]:
bacteria.loc['Actinobacteria']

bacteria_counts    569
other_feature      234
Name: Actinobacteria, dtype: int64

We could have also used "positional indexing," even though the indices are strings. The difference is that `.loc` returns a `Series` because we selected a single label, while `[2:3]` returns a `DataFrame` because we selected a range of positions.

In [30]:
bacteria[2:3]

bacteria_counts  other_feature
Actinobacteria              569            234

Let's return to the *unemployment* data to show how `.iloc` is used, since *unemployment* has integer indices. To select specific rows, we can do the following.

In [31]:
unemployment.iloc[0:4]

country seasonality  year_month  unemployment  unemployment_rate
0      at         nsa     1993.01        171000                4.5
1      at         nsa     1993.02        175000                4.6
2      at         nsa     1993.03        166000                4.4
3      at         nsa     1993.04        157000                4.1

In [32]:
unemployment.iloc[[1, 5, 6, 22]]

country seasonality  year_month  unemployment  unemployment_rate
1       at         nsa     1993.02        175000                4.6
5       at         nsa     1993.06        134000                3.5
6       at         nsa     1993.07        128000                3.4
22      at         nsa     1994.11        148000                3.9

We can also select a range of rows and specify the step value.

In [33]:
unemployment.iloc[25:50:5]

country seasonality  year_month  unemployment  unemployment_rate
25      at         nsa     1995.02        174000                4.5
30      at         nsa     1995.07        123000                3.3
35      at         nsa     1995.12        175000                4.7
40      at         nsa     1996.05        159000                4.3
45      at         nsa     1996.10        146000                3.9

### Selecting a Specific Value

Both `.loc` and `.iloc` can be used to select a particular value if they are given two arguments. The first argument is the row name (when using `.loc`) or the row index number (when using `.iloc`), while the second argument is the column name or index number.

Using `loc`, we can select "Bacteroidetes" and "bacteria_counts" to get the count of Bacteroidetes.

In [34]:
bacteria

bacteria_counts  other_feature
Firmicutes                  632            438
Proteobacteria             1638            833
Actinobacteria              569            234
Bacteroidetes               115            298

In [35]:
bacteria.loc['Bacteroidetes']['bacteria_counts']

115

In [36]:
# This is the same as above
bacteria.iloc[3][0]

115

In [37]:
# This the same as above
bacteria.iloc[3]['bacteria_counts']

115

Or, for the *unemployment* data:

In [38]:
# The year_month in the first row
unemployment.iloc[0,2]

1993.01

In [39]:
# This the same as above
unemployment.iloc[0][2]

1993.01

### Selecting Multiple Rows and Columns

Both `.loc` and `.iloc` can be used to select subsets of rows and columns at the same time if they are given lists (and/or slices for `.iloc`) as their two arguments. 

Using `.iloc` on the *unemployment* DataFrame, select the rows starting at row 2 and ending at row 5, and the 0th, 2nd, and 3rd columns.

In [40]:
unemployment.iloc[2:6,[0,2,3]]

country  year_month  unemployment
2      at     1993.03        166000
3      at     1993.04        157000
4      at     1993.05        147000
5      at     1993.06        134000

The same selection can be achieved by using the `.loc` operator and listing the column names.

In [41]:
# The same as above
unemployment.loc[2:6,['country', 'year_month', 'unemployment']]

country  year_month  unemployment
2      at     1993.03        166000
3      at     1993.04        157000
4      at     1993.05        147000
5      at     1993.06        134000
6      at     1993.07        128000

We can also display values from a DataFrame that satisfy certain criteria using conditional expressions, such as <, >, ==, !=, etc. 

One example is shown below where only the rows that have an unemployment rate greater than 15 are shown. 

In [42]:
unemployment[unemployment['unemployment_rate'] > 15].head(10)

country seasonality  year_month  unemployment  unemployment_rate
1717      bg         nsa     2000.02        523000               15.4
1718      bg         nsa     2000.03        547000               16.0
1719      bg         nsa     2000.04        560000               16.3
1720      bg         nsa     2000.05        561000               16.3
1721      bg         nsa     2000.06        554000               16.2
1722      bg         nsa     2000.07        558000               16.3
1723      bg         nsa     2000.08        569000               16.7
1724      bg         nsa     2000.09        574000               16.8
1725      bg         nsa     2000.10        583000               17.1
1726      bg         nsa     2000.11        597000               17.5

# 10.4 Creating New Columns, Reordering <a id="section4"/>

Since the `year_month` column is not correct, let's try to split it into two separate columns for years and months. 

In the previous section we saw that the data type in this column is `float64`. We'll first extract the year using the `.astype()` method. This allows for type casting, i.e., using `.astype(int)` we will convert the floating point values into integer numbers (by truncating the decimals).

The new column *year* will be added on the right of the DataFrame.

In [43]:
unemployment['year'] = unemployment['year_month'].astype(int)
unemployment.head()

country seasonality  year_month  unemployment  unemployment_rate  year
0      at         nsa     1993.01        171000                4.5  1993
1      at         nsa     1993.02        175000                4.6  1993
2      at         nsa     1993.03        166000                4.4  1993
3      at         nsa     1993.04        157000                4.1  1993
4      at         nsa     1993.05        147000                3.9  1993

Next, let's create a new column *month*. We will subtract the *year* value from *year_month* to get the decimal portion of the value, and multiply the result by 100 and convert to `int`. Because of the truncating that occurs when casting to `int`, we first need to round the values to the nearest whole number using `round()`. 

In [44]:
unemployment['month'] = ((unemployment['year_month'] - unemployment['year']) * 100).round(0).astype(int)
unemployment.head(12)

country seasonality  year_month  unemployment  unemployment_rate  year  \
0       at         nsa     1993.01        171000                4.5  1993   
1       at         nsa     1993.02        175000                4.6  1993   
2       at         nsa     1993.03        166000                4.4  1993   
3       at         nsa     1993.04        157000                4.1  1993   
4       at         nsa     1993.05        147000                3.9  1993   
5       at         nsa     1993.06        134000                3.5  1993   
6       at         nsa     1993.07        128000                3.4  1993   
7       at         nsa     1993.08        130000                3.4  1993   
8       at         nsa     1993.09        132000                3.5  1993   
9       at         nsa     1993.10        141000                3.7  1993   
10      at         nsa     1993.11        156000                4.1  1993   
11      at         nsa     1993.12        169000                4.4  1993   

    month  
0       1  
1       2  
2       3  
3       4  
4       5  
5       6  
6       7  
7       8  
8       9  
9      10  
10     11  
11     12

Now, let's try to reorder the newly created *year* and *month* columns in the DataFrame. For this, we use the square brackets notation again, passing in a list of column names in the order we would like to see them.

In [45]:
unemployment = unemployment[['country', 'seasonality',
                             'year_month', 'year', 'month',
                             'unemployment', 'unemployment_rate']]
unemployment.head(10)

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   
5      at         nsa     1993.06  1993      6        134000   
6      at         nsa     1993.07  1993      7        128000   
7      at         nsa     1993.08  1993      8        130000   
8      at         nsa     1993.09  1993      9        132000   
9      at         nsa     1993.10  1993     10        141000   

   unemployment_rate  
0                4.5  
1                4.6  
2                4.4  
3                4.1  
4                3.9  
5                3.5  
6                3.4  
7                3.4  
8                3.5  
9                3.7

# 10.5 Merging <a id="section5"/>

If we examine the `unemployment` DataFrame we can notice that we don't exactly know what the values in the `country` column refer to. We can fix that by getting the country names from the `countries` DataFrame that we imported earlier. 

We can see in the `countries` data that *at* stands for Austria. The data even provides the country names in three different languages.

In [46]:
countries.head()

country google_country_code country_group   name_en   name_fr     name_de  \
0      at                  AT            eu   Austria  Autriche  Österreich   
1      be                  BE            eu   Belgium  Belgique     Belgien   
2      bg                  BG            eu  Bulgaria  Bulgarie   Bulgarien   
3      hr                  HR        non-eu   Croatia   Croatie    Kroatien   
4      cy                  CY            eu    Cyprus    Chypre      Zypern   

    latitude  longitude  
0  47.696554  13.345980  
1  50.501045   4.476674  
2  42.725674  25.482322  
3  44.746643  15.340844  
4  35.129141  33.428682

Because the data we need is stored in two separate files, we will first merge the two DataFrames. The `country` column is shown in both DataFrames, so it is a good option for joining the data. However, we don't need all of the columns in the `countries` DataFrame, and therefore, we will create a new DataFrame. To select certain columns to retain, we can use the bracket notation that we used earlier to reorder the columns.

In [47]:
country_names = countries[['country', 'country_group', 'name_en']]

In [48]:
country_names.head(5)

country country_group   name_en
0      at            eu   Austria
1      be            eu   Belgium
2      bg            eu  Bulgaria
3      hr        non-eu   Croatia
4      cy            eu    Cyprus

`pandas` include an easy-to-use `merge` function, which has the following syntax.

```
pd.merge(first_file, second_file, on=['column_name'])
```


In [49]:
unemployment = pd.merge(unemployment, country_names, on=['country'])
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate country_group  name_en  
0                4.5            eu  Austria  
1                4.6            eu  Austria  
2                4.4            eu  Austria  
3                4.1            eu  Austria  
4                3.9            eu  Austria

If we want to merge two files using multiple columns that exist in both files, we can pass a list of column names to the `on` parameter.

For more information on merging, check the pandas [documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html#database-style-dataframe-joining-merging).

# 10.6 Calculating Unique and Missing Values <a id="section6"/>

In the *unemployment* data, we might want to know what countries we have data for. To extract this information, we can use the `.unique()` method. Note that the countries are listed in the right-most column `name-en` so we will use it to find the unique elements in that column.

In [50]:
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate country_group  name_en  
0                4.5            eu  Austria  
1                4.6            eu  Austria  
2                4.4            eu  Austria  
3                4.1            eu  Austria  
4                3.9            eu  Austria

In [51]:
unemployment.name_en.unique()

array(['Austria', 'Belgium', 'Bulgaria', 'Cyprus', 'Czech Republic',
       'Germany (including  former GDR from 1991)', 'Denmark', 'Estonia',
       'Spain', 'Finland', 'France', 'Greece', 'Croatia', 'Hungary',
       'Ireland', 'Italy', 'Lithuania', 'Luxembourg', 'Latvia', 'Malta',
       'Netherlands', 'Norway', 'Poland', 'Portugal', 'Romania', 'Sweden',
       'Slovenia', 'Slovakia', 'Turkey', 'United Kingdom'], dtype=object)

To get a count of the **number of unique countries,** we can use the `.nunique()` method.

In [52]:
unemployment.name_en.nunique()

30

Or, we can also use `len()` to get the number of items in the array

In [53]:
len(unemployment.name_en.unique())

30

It might be interesting to know **how many observations** we have per country. `pandas` have a method called `.value_counts()` that returns the counts for the unique values in a column.

In [54]:
unemployment['name_en'].value_counts()

Netherlands                                  1008
Sweden                                       1008
Belgium                                      1008
Luxembourg                                   1008
Portugal                                     1008
France                                       1008
Spain                                        1008
Ireland                                      1008
Denmark                                      1008
United Kingdom                               1002
Italy                                         924
Finland                                       828
Norway                                        786
Austria                                       648
Hungary                                       576
Slovakia                                      576
Bulgaria                                      576
Slovenia                                      576
Poland                                        576
Malta                                         576


By default, the output is sorted by values. If we would like it sorted by index (or, country name in this case), append the `.sort_index()` method.

In [55]:
unemployment['name_en'].value_counts().sort_index()

Austria                                       648
Belgium                                      1008
Bulgaria                                      576
Croatia                                       324
Cyprus                                        396
Czech Republic                                468
Denmark                                      1008
Estonia                                       387
Finland                                       828
France                                       1008
Germany (including  former GDR from 1991)     504
Greece                                        450
Hungary                                       576
Ireland                                      1008
Italy                                         924
Latvia                                        459
Lithuania                                     459
Luxembourg                                   1008
Malta                                         576
Netherlands                                  1008


As we noticed earlier, there are missing values in the `unemployment_rate` column. To find out **how many unemployment rate values are missing** we will use the `.isnull()` method, which returns a corresponding boolean value for each missing entry in the unemployment_rate column. As we know, in Python `True` is equivalent to 1 and `False` is equivalent to 0. Thus, when we add the result with `.sum()`, we get a count for the total number of missing values.

In [56]:
unemployment['unemployment_rate'].isnull().sum()

945

### GroupBy

What if we would like to know how many missing values exist at the *country* level? We can take the main part of what we had above and create a new column in the `DataFrame`. This is the last column, in which `False` means that the value is not missing.

In [57]:
unemployment['unemployment_rate_null'] = unemployment['unemployment_rate'].isnull()
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate country_group  name_en  unemployment_rate_null  
0                4.5            eu  Austria                   False  
1                4.6            eu  Austria                   False  
2                4.4            eu  Austria                   False  
3                4.1            eu  Austria                   False  
4                3.9            eu  Austria                   False

To count the **number of missing values for each country,** we can use the `.groupby()` method. It groups the data by the country `name_en` column included in the parentheses, and `.sum()` operation is performed over the `unemployment_rate_null` column.

In [58]:
unemployment.groupby('name_en')['unemployment_rate_null'].sum()

name_en
Austria                                        0.0
Belgium                                        0.0
Bulgaria                                     180.0
Croatia                                      216.0
Cyprus                                         0.0
Czech Republic                                 0.0
Denmark                                        0.0
Estonia                                        0.0
Finland                                        0.0
France                                         0.0
Germany (including  former GDR from 1991)      0.0
Greece                                         0.0
Hungary                                       36.0
Ireland                                        0.0
Italy                                          0.0
Latvia                                         0.0
Lithuania                                      0.0
Luxembourg                                     0.0
Malta                                        180.0
Netherlands            

# 10.7 Exporting A DataFrame to csv <a id="section7"/>

If we wanted to save the last DataFrame as a .csv file, we can use the `.to_csv()` method.

In [59]:
unemployment.to_csv('data/unemployment_missing.csv')

The file is saved in the `data` directory. 

By default, this method writes the indices. We probably don't want the index row, and we can set `index` to `False`. We can also specify the type of delimiter that we want to use, such as commas `(,)`, pipes (`|`), semicolons (`;`), tabs (`\t`), etc.

In [60]:
unemployment.to_csv('data/unemployment_missing.csv', index=False, sep=',')

Now that we have the missing values saved, we can *drop* the last column `unemployment_rate_null` that we added to `unemployment`, as explained below.

In [61]:
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate country_group  name_en  unemployment_rate_null  
0                4.5            eu  Austria                   False  
1                4.6            eu  Austria                   False  
2                4.4            eu  Austria                   False  
3                4.1            eu  Austria                   False  
4                3.9            eu  Austria                   False

In [62]:
unemployment.drop('unemployment_rate_null', axis=1, inplace=True)
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate country_group  name_en  
0                4.5            eu  Austria  
1                4.6            eu  Austria  
2                4.4            eu  Austria  
3                4.1            eu  Austria  
4                3.9            eu  Austria

It's important to specify the `axis` parameter, where `axis=1` refers to columns (`axis=0` refers to rows). The parameter `inplace=True` modifies the actual DataFrame rather than returning a new DataFrame.

# 10.8 Dealing With Missing Values: Boolean Indexing <a id="section8"/>

There are two main options for dealing with missing values:

* Fill the missing values with some other values.
* Do not use observations with missing values.
    
Here we will adopt the second approach and **exclude missing values** from our primary analyses. Instead of just getting rid of that data, it might make sense to store it in a separate DataFrame. This way, we could answer questions such as, "do missing values occur during certain months (or years) more frequently?" 

To do this, we will use *boolean indexing* for filtering data.

Recall that `unemployment['unemployment_rate'].isnull()` produces an array of boolean values. We used this previously when counting the number of missing values. This is shown in the next cell.

In [63]:
unemployment['unemployment_rate'].isnull()[:10]

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9    False
Name: unemployment_rate, dtype: bool

To create `unemployment_rate_missing`, we will index `unemployment` with the array above. This returns only the rows where the value in the array is `True`.

In [64]:
unemployment_rate_missing = unemployment[unemployment['unemployment_rate'].isnull()]
unemployment_rate_missing.head()

country seasonality  year_month  year  month  unemployment  \
1656      bg         nsa     1995.01  1995      1        391000   
1657      bg         nsa     1995.02  1995      2        387000   
1658      bg         nsa     1995.03  1995      3        378000   
1659      bg         nsa     1995.04  1995      4        365000   
1660      bg         nsa     1995.05  1995      5        346000   

      unemployment_rate country_group   name_en  
1656                NaN            eu  Bulgaria  
1657                NaN            eu  Bulgaria  
1658                NaN            eu  Bulgaria  
1659                NaN            eu  Bulgaria  
1660                NaN            eu  Bulgaria

It is also possible to specify multiple conditions using the `&` operator, but each condition needs to be inside of parentheses. 

Now, to remove the missing data in `unemployment`, we can use the `.dropna()` method. This method drops all observations for which `unemployment_rate == NaN`.

In [65]:
unemployment.dropna(subset=['unemployment_rate'], inplace=True)
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate country_group  name_en  
0                4.5            eu  Austria  
1                4.6            eu  Austria  
2                4.4            eu  Austria  
3                4.1            eu  Austria  
4                3.9            eu  Austria

### Sorting Values

If we want to know what the highest unemployment rates were, we can use the `.sort_values()` method to *sort the data.*

This cell sorted the data in *descending* order, and printed the first 10 rows.

In [66]:
unemployment.sort_values('unemployment_rate', ascending=False)[:10]

country seasonality  year_month  year  month  unemployment  \
15526      pl         nsa     2004.02  2004      2       3531000   
15525      pl         nsa     2004.01  2004      1       3520000   
15514      pl         nsa     2003.02  2003      2       3460000   
5663       es          sa     2010.09  2010      9       4773000   
15527      pl         nsa     2004.03  2004      3       3475000   
5999       es       trend     2010.09  2010      9       4769000   
6000       es       trend     2010.10  2010     10       4760000   
15513      pl         nsa     2003.01  2003      1       3466000   
5664       es          sa     2010.10  2010     10       4758000   
5998       es       trend     2010.08  2010      8       4742000   

       unemployment_rate country_group name_en  
15526               20.9            eu  Poland  
15525               20.7            eu  Poland  
15514               20.7            eu  Poland  
5663                20.6            eu   Spain  
15527               20.6            eu  Poland  
5999                20.6            eu   Spain  
6000                20.6            eu   Spain  
15513               20.6            eu  Poland  
5664                20.6            eu   Spain  
5998                20.5            eu   Spain

# References <a id="section9"/>

1. Introduction to Pandas, Python Data Wrangling by D-Lab at UC Berkley, available at: [https://github.com/dlab-berkeley/introduction-to-pandas](https://github.com/dlab-berkeley/introduction-to-pandas).
2. Pandas documentation, available at [https://pandas.pydata.org/pandas-docs/stable/](https://pandas.pydata.org/pandas-docs/stable/).